x/gamm/twap/logic.go

```diff
@@ -0,0 +1,149 @@
+package twap
+
+import (
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/osmosis-labs/osmosis/v10/x/gamm/twap/types"
+)
+
+func (k Keeper) afterCreatePool(ctx sdk.Context, poolId uint64) error {
+	denoms, err := k.ammkeeper.GetPoolDenoms(ctx, poolId)
+	denomPairs0, denomPairs1 := types.GetAllUniqueDenomPairs(denoms)
+	for i := 0; i < len(denomPairs0); i++ {
+		record, err := types.NewTwapRecord(k.ammkeeper, ctx, poolId, denomPairs0[i], denomPairs1[i])
+		// err should be impossible given GetAllUniqueDenomPairs guarantees
+		if err != nil {
+			return err
+		}
+		k.storeNewRecord(ctx, record)
+	}
+	k.trackChangedPool(ctx, poolId)
+	return err
+}
+
+func (k Keeper) endBlock(ctx sdk.Context) {
+	changedPoolIds := k.getChangedPools(ctx)
+	for _, id := range changedPoolIds {
+		err := k.updateRecords(ctx, id)
+		if err != nil {
+			panic(err)
+		}
+	}
+	// 'altered pool ids' gets automatically cleared by being a transient store
```

Comment:
nit: I would just add this comment as part of the godoc.

Reply:
oh yeah lol, that is an awkward dangling comment haha